src/asciidoctorWebViewConverter.ts

```diff
@@ -58,20 +58,33 @@ const previewStrings = {
   const highlightjsInlineScriptHash = 'sha256-ZrDBcrmObbqhVV/Mag2fT/y08UJGejdW7UWyEsi4DXw='
+  let cspRule = '';
   switch (securityLevel) {
     case AsciidocPreviewSecurityLevel.AllowInsecureContent:
-      return `<meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src 'self' ${rule} http: https: data:; media-src 'self' ${rule} http: https: data:; script-src 'nonce-${nonce}' '${highlightjsInlineScriptHash}'; style-src 'self' ${rule} 'unsafe-inline' http: https: data:; font-src 'self' ${rule} http: https: data:;">`
+      cspRule += `default-src 'none'; img-src 'self' ${rule} http: https: data:; media-src 'self' ${rule} http: https: data:; script-src 'nonce-${nonce}' '${highlightjsInlineScriptHash}'; style-src 'self' ${rule} 'unsafe-inline' http: https: data:; font-src 'self' ${rule} http: https: data:;`
+      break
 
     case AsciidocPreviewSecurityLevel.AllowInsecureLocalContent:
-      return `<meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src 'self' ${rule} https: data: http://localhost:* http://127.0.0.1:*; media-src 'self' ${rule} https: data: http://localhost:* http://127.0.0.1:*; script-src 'nonce-${nonce}' '${highlightjsInlineScriptHash}'; style-src 'self' ${rule} 'unsafe-inline' https: data: http://localhost:* http://127.0.0.1:*; font-src 'self' ${rule} https: data: http://localhost:* http://127.0.0.1:*;">`
+      cspRule += `default-src 'none'; img-src 'self' ${rule} https: data: http://localhost:* http://127.0.0.1:*; media-src 'self' ${rule} https: data: http://localhost:* http://127.0.0.1:*; script-src 'nonce-${nonce}' '${highlightjsInlineScriptHash}'; style-src 'self' ${rule} 'unsafe-inline' https: data: http://localhost:* http://127.0.0.1:*; font-src 'self' ${rule} https: data: http://localhost:* http://127.0.0.1:*;`
+      break
 
     case AsciidocPreviewSecurityLevel.AllowScriptsAndAllContent:
-      return '<meta http-equiv="Content-Security-Policy" content="">'
+      cspRule = ''
+      break
 
     case AsciidocPreviewSecurityLevel.Strict:
-    default:
-      return `<meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src 'self' ${rule} https: data:; media-src 'self' ${rule} https: data:; script-src 'nonce-${nonce}' '${highlightjsInlineScriptHash}'; style-src 'self' ${rule} 'unsafe-inline' https: data:; font-src 'self' ${rule} https: data:;">`
+      default:
+      cspRule += `default-src 'none'; img-src 'self' ${rule} https: data:; media-src 'self' ${rule} https: data:; script-src 'nonce-${nonce}' '${highlightjsInlineScriptHash}'; style-src 'self' ${rule} 'unsafe-inline' https: data:; font-src 'self' ${rule} https: data:;`
+      break
   }
+     // Check if the enableKroki configuration is set to true
+  if (vscode.workspace.getConfiguration('asciidoc.extensions', null).get('enableKroki')) {
+    // Append the kroki-server-url to the CSP rule
+    cspRule += `; img-src 'self' ${kroki-server-url};`
```

Comment:
`${kroki-server-url}` is undefined, you need to pass the value

Reply:
Oof, I need to think about eliminating duplication of the getConfiguration code here too. Right now, I am thinking in the parser file:
```
// Check if the enableKroki configuration is set to true
  const enableKroki = vscode.workspace.getConfiguration('asciidoc.extensions', null).get('enableKroki');
  if (enableKroki) {
    // Check if the krokiServerUrl property is set
    if (vscode.workspace.getConfiguration('asciidoc.extensions', null).has('krokiServerUrl')) {
      // Get the krokiServerUrl property
      const krokiServerUrl = vscode.workspace.getConfiguration('asciidoc.extensions', null).get('krokiServerUrl');

      // Append the kroki-server-url to the CSP rule
      cspRule += `; img-src 'self' ${krokiServerUrl};`
    } 
```

And then change the if statement to: 

```
// Check if the krokiServerUrl property is set
if (vscode.workspace.getConfiguration('asciidoc.extensions', null).has('krokiServerUrl')) {
  // Get the krokiServerUrl property
  const krokiServerUrl = vscode.workspace.getConfiguration('asciidoc.extensions', null).get('krokiServerUrl');

  // Check if the enableKroki configuration is set to true
  const enableKroki = vscode.workspace.getConfiguration('asciidoc.extensions', null).get('enableKroki');
  if (enableKroki) {
    // Append the kroki-server-url to the CSP rule
    cspRule += `; img-src 'self' ${krokiServerUrl};`
  }
}
```

Reply:
`kroki-server-url` is an AsciiDoc attribute not a VS Code configuration, that's why you need to use the `AsciidocPreviewConfiguration` class.
The default value is `https://kroki.io` if the value is not explicitly defined.

Reply:
Okay so I need tp add kroki-server-url as an AsciiDoc attribute to the AsciidocPreviewConfiguration class
```
export class AsciidocPreviewConfiguration {
  public static getForResource (resource: vscode.Uri) {
    return new AsciidocPreviewConfiguration(resource)
  }

  // ... other properties and methods

  public readonly krokiServerUrl: string
  
  private constructor (resource: vscode.Uri) {
    // ... other property assignments

    const asciidocConfig = vscode.workspace.getConfiguration('asciidoc', resource)
    this.krokiServerUrl = asciidocConfig.get<string>('preview.krokiServerUrl', 'https://kroki.io')
  }
}
```
Then the value of kroki-server-url should pass to the AsciidocPreviewConfiguration class. Then come back to the cspRules here and add:
```
// Check if the enableKroki configuration is set to true
const enableKroki = vscode.workspace.getConfiguration('asciidoc.extensions', null).get('enableKroki')
if (enableKroki) {
  // Get the value of the krokiServerUrl property from the AsciidocPreviewConfiguration class
  const config = AsciidocPreviewConfiguration.getForResource(resource)
  const krokiServerUrl = config.krokiServerUrl

  // Append the krokiServerUrl to the CSP rule
  cspRule += `; img-src 'self' ${krokiServerUrl};`
}
```

Reply:
Not exactly, we need to load the AsciiDoc document and call `getAttribute()`: 

https://github.com/asciidoctor/asciidoctor-vscode/blob/117449b31216d9492e77d26b147610ed51bfee88/src/asciidocParser.ts#L203

You can have a look at: https://github.com/Mogztter/asciidoctor-kroki/blob/059510b500ad9e39733c27ba88b563736fc17ade/src/kroki-client.js#L76

Then, we need to pass this value to `AsciidoctorWebViewConverter` in order to compute the CSP correctly.

Reply:
Do I need to load the Asciidoc document in the private constructor method or somewhere else within the 'AsciidocPreviewConfiguration' class? Confused about where to implement loading the document because I was initially under the impression that the AsciidocPreviewConfirguration class just pulled from the users vscode workspace? I can see that kroki-client you sent me now is doing something else on its end besides just using the users vscode workspace.

Reply:
I think we should load the Asciidoctor document using the `header_only` before we instantiate the `AsciidoctorWebViewConverter` class in:

https://github.com/asciidoctor/asciidoctor-vscode/blob/1e07c19864bc228661455f476a2d4e329c5fba45/src/asciidocParser.ts#L148-L157

```js
const asciidoctorDocument = processor.load(text, { header_only: true })
const krokiServerUrl = document.getAttribute('kroki-server-url') || 'https://kroki.io'

new AsciidoctorWebViewConverter(/* pass the krokiServerUrl here */)
```

As a side note, the AsciidoctorWebViewConverter constructor has quite a few arguments and we should probably refine it but we can do that later. For now, just add a new constructor argument named `krokiServerUrl`.

Does it make sense?

Reply:
Okay, so to summarize before I make changes to my commit. I will be loading the Asciidoctor document using the `header_only` before we instantiate the `AsciidoctorWebViewConverter` in `asciidocParser.ts`, as shown by:

```
// load the asciidoc header only to get kroki-server-url attribute
const asciidoctorDocument = processor.load(text, { header_only: true })
const krokiServerUrl = document.getAttribute('kroki-server-url') || 'https://kroki.io'

const asciidoctorWebViewConverter = new AsciidoctorWebViewConverter( 
   doc, 
   editor, 
   cspArbiter.getSecurityLevelForResource(doc.uri), 
   cspArbiter.shouldDisableSecurityWarnings(), 
   this.contributionProvider.contributions, 
   previewConfigurationManager.loadAndCacheConfiguration(doc.uri), 
   antoraDocumentContext, 
   line,
   krokiServerUrl
 ) 
```
Then krokiServerUrl should be defined,
``

Reply:
Yes that's correct.

> Do I also need to add the krokiServerUrl here as initial data? -

I don't think so.